src/std_extensions/quantum.rs

```diff
@@ -1,19 +1,190 @@
 //! Basic HUGR quantum operations
 
+use std::cmp::max;
+use std::f64::consts::TAU;
+use std::num::NonZeroU64;
+
 use smol_str::SmolStr;
 
-use crate::extension::prelude::{BOOL_T, QB_T};
+use crate::extension::prelude::{BOOL_T, ERROR_TYPE, QB_T};
 use crate::extension::{ExtensionId, SignatureError};
 use crate::std_extensions::arithmetic::float_types::FLOAT64_TYPE;
 use crate::type_row;
-use crate::types::type_param::TypeArg;
-use crate::types::FunctionType;
+use crate::types::type_param::{TypeArg, TypeArgError, TypeParam};
+use crate::types::{ConstTypeError, CustomCheckFailure, CustomType, FunctionType, Type, TypeBound};
+use crate::utils::collect_array;
+use crate::values::CustomConst;
 use crate::Extension;
 
 use lazy_static::lazy_static;
 
 /// The extension identifier.
 pub const EXTENSION_ID: ExtensionId = ExtensionId::new_unchecked("quantum");
+
+/// Identifier for the angle type.
+const ANGLE_TYPE_ID: SmolStr = SmolStr::new_inline("angle");
+
+fn angle_custom_type(log_denom_arg: TypeArg) -> CustomType {
+    CustomType::new(ANGLE_TYPE_ID, [log_denom_arg], EXTENSION_ID, TypeBound::Eq)
+}
+
+/// Angle type with a given log-denominator (specified by the TypeArg).
+///
+/// This type is capable of representing angles that are multiples of 2π / 2^N where N is the
+/// log-denominator.
+pub(super) fn angle_type(log_denom_arg: TypeArg) -> Type {
+    Type::new_extension(angle_custom_type(log_denom_arg))
+}
+
+/// The smallest forbidden log-denominator.
+pub const LOG_DENOM_BOUND: u8 = 54;
+
+const fn is_valid_log_denom(n: u8) -> bool {
+    n < LOG_DENOM_BOUND
+}
+
+/// Type parameter for the log-denominator of an angle.
+// SAFETY: unsafe block should be ok as the value is definitely not zero.
+#[allow(clippy::assertions_on_constants)]
+pub const LOG_DENOM_TYPE_PARAM: TypeParam = TypeParam::bounded_nat(unsafe {
```

Comment:
Whoa! `new_unchecked` needs `unsafe`, I see. Hmmmm. Ick. (I think? Or should we make our `new_unchecked`s also need unsafe?)

The following works without an `unsafe`, but you can keep the assert:
`NonZeroU64::MIN.saturating_add((LOG_WIDTH_BOUND-1) as u64)`

or even
`NonZeroU64::MIN.saturating_add((LOG_WIDTH_BOUND as u64)-NonZeroU64::MIN.get())`

Ymmv :)

Reply:
Nice, we can probably do this in the integer extension code too where there is a similar unsafe block.